File: 111-minimum-depth-of-binary-tree/111-minimum-depth-of-binary-tree.py
from collections import deque
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        queue = deque()
        queue.append(root)
        count = 0
        if not root:
            return 0
        while(queue):
            size = len(queue)
            count += 1
            for i in range(size):
                cur = queue.popleft()
                if not cur.left and not cur.right:
                    return count
                if cur.left:
                    queue.append(cur.left)
                if cur.right:
                    queue.append(cur.right)
        

File: 111-minimum-depth-of-binary-tree/111-minimum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        //@input root of a binary tree
        //@output the minimum depth the number of nodes from root down to the nearest leaf
        //bfs/dfs
        LinkedList<TreeNode> queue = new LinkedList<>();
        if(root == null)return 0;
        queue.addLast(root);
        int minDepth = 0;
        while(!queue.isEmpty()){
            int size = queue.size();
            minDepth += 1;
            for(int i = 0; i < size; i++){
                TreeNode popped = queue.pollFirst();
                if(popped.left == null && popped.right == null){
                    return minDepth;
                }
                if(popped.left != null){
                    queue.addLast(popped.left);
                }
                if(popped.right != null){
                    queue.addLast(popped.right);
                }
                
            }
        }
        return minDepth;
        
        
        
    }
}